src/parquet/column/writer.h

```diff
@@ -0,0 +1,206 @@
+
+  // Write multiple definition levels
+  void WriteDefinitionLevels(int64_t num_levels, int16_t* levels);
+
+  // Write multiple repetition levels
+  void WriteRepetitionLevels(int64_t num_levels, int16_t* levels);
+
+  std::shared_ptr<Buffer> RleEncodeLevels(const std::shared_ptr<Buffer>& buffer,
+      int16_t max_level);
+
+  const ColumnDescriptor* descr_;
+
+  std::unique_ptr<PageWriter> pager_;
+
+  // The number of rows that should be written in this column chunk.
+  int64_t expected_rows_;
+
+  LevelEncoder level_encoder_;
+
+  MemoryAllocator* allocator_;
+
+  // The total number of values stored in the data page. This is the maximum of
+  // the number of encoded definition levels or encoded values. For
+  // non-repeated, required columns, this is equal to the number of encoded
+  // values. For repeated or optional values, there may be fewer data values
+  // than levels, and this tells you how many encoded levels there are in that
+  // case.
+  int num_buffered_values_;
+
+  // The total number of stored values. For repeated or optional values, this
+  // number may be lower than num_buffered_values_.
+  int num_buffered_encoded_values_;
+
+  // Total number of rows written with this ColumnWriter
+  int num_rows_;
+
+  int total_bytes_written_;
+
+  std::unique_ptr<InMemoryOutputStream> definition_levels_sink_;
+  std::unique_ptr<InMemoryOutputStream> repetition_levels_sink_;
+  std::unique_ptr<InMemoryOutputStream> values_sink_;
+
+ private:
+  void InitSinks();
+};
+
+// API to read values from a single column. This is the main client facing API.
+template <int TYPE>
+class TypedColumnWriter : public ColumnWriter {
+ public:
+  typedef typename type_traits<TYPE>::value_type T;
+
+  TypedColumnWriter(const ColumnDescriptor* schema,
+      std::unique_ptr<PageWriter> pager, int64_t expected_rows,
+      MemoryAllocator* allocator = default_allocator());
+
+  // Write a batch of repetition levels, definition levels, and values to the
+  // column.
+  void WriteBatch(int64_t num_values, int16_t* def_levels, int16_t* rep_levels,
+      T* values);
+
+ private:
+  typedef Encoder<TYPE> EncoderType;
+
+  // Write values to a temporary buffer before they are encoded into pages
+  void WriteValues(int64_t num_values, T* values);
+
+  // Map of encoding type to the respective encoder object. For example, a
+  // column chunk's data pages may include both dictionary-encoded and
+  // plain-encoded data.
+  std::unordered_map<int, std::shared_ptr<EncoderType> > encoders_;
+
+  void ConfigureDictionary(const DictionaryPage* page);
+
+  std::unique_ptr<EncoderType> current_encoder_;
+};
+
+// TODO(PARQUET-591): This is just chosen at random, we should make better estimates.
+// See also: parquet-column/../column/impl/ColumnWriteStoreV2.java:sizeCheck
+const int64_t PAGE_VALUE_COUNT = 1000;
+
+template <int TYPE>
+inline void TypedColumnWriter<TYPE>::WriteBatch(int64_t num_values, int16_t* def_levels,
+    int16_t* rep_levels, T* values) {
```

Comment:
On some contemplation, I'm curious whether the API should be one page at a time versus accumulate-then-write. This leaves the user to be responsible for page delimiting.